File: regions/system/library/swift/websocket/Sources/K1s0WebSocket/WsClient.swift
import Foundation

public enum WsError: Error, Sendable {
    case notConnected
    case noMessagesAvailable
    case alreadyConnected
    case connectionFailed(reason: String)
}

public protocol WsClient: Sendable {
    func connect() async throws
    func disconnect() async throws
    func send(_ message: WsMessage) async throws
    func receive() async throws -> WsMessage
    var state: ConnectionState { get async }
}

public actor InMemoryWsClient: WsClient {
    private var _state: ConnectionState = .disconnected
    private var receiveBuffer: [WsMessage] = []
    private var sentMessages: [WsMessage] = []
    private let config: WsConfig

    public init(config: WsConfig = .defaults) {
        self.config = config
    }

    public var state: ConnectionState {
        _state
    }

    public func connect() async throws {
        guard _state == .disconnected || _state == .reconnecting else {
            throw WsError.alreadyConnected
        }
        _state = .connecting
        _state = .connected
    }

    public func disconnect() async throws {
        _state = .closing
        _state = .disconnected
        receiveBuffer.removeAll()
    }

    public func send(_ message: WsMessage) async throws {
        guard _state == .connected else {
            throw WsError.notConnected
        }
        sentMessages.append(message)
    }

    public func receive() async throws -> WsMessage {
        guard _state == .connected else {
            throw WsError.notConnected
        }
        guard !receiveBuffer.isEmpty else {
            throw WsError.noMessagesAvailable
        }
        return receiveBuffer.removeFirst()
    }

    public func injectMessage(_ msg: WsMessage) {
        receiveBuffer.append(msg)
    }

    public func getSentMessages() -> [WsMessage] {
        sentMessages
    }
}

File: regions/system/library/dart/websocket/lib/src/ws_client.dart
import 'dart:async';
import 'dart:collection';

import 'ws_message.dart';
import 'connection_state.dart';

abstract class WsClient {
  Future<void> connect();
  Future<void> disconnect();
  Future<void> send(WsMessage message);
  Future<WsMessage> receive();
  ConnectionState get state;
}

class InMemoryWsClient implements WsClient {
  final _receiveQueue = Queue<WsMessage>();
  final _sentMessages = <WsMessage>[];
  ConnectionState _state = ConnectionState.disconnected;

  @override
  ConnectionState get state => _state;

  List<WsMessage> get sentMessages => List.unmodifiable(_sentMessages);

  void injectMessage(WsMessage msg) {
    _receiveQueue.add(msg);
  }

  @override
  Future<void> connect() async {
    if (_state == ConnectionState.connected) {
      throw StateError('Already connected');
    }
    _state = ConnectionState.connecting;
    _state = ConnectionState.connected;
  }

  @override
  Future<void> disconnect() async {
    if (_state == ConnectionState.disconnected) {
      throw StateError('Not connected');
    }
    _state = ConnectionState.closing;
    _state = ConnectionState.disconnected;
  }

  @override
  Future<void> send(WsMessage message) async {
    if (_state != ConnectionState.connected) {
      throw StateError('Cannot send message while $state');
    }
    _sentMessages.add(message);
  }

  @override
  Future<WsMessage> receive() async {
    if (_state != ConnectionState.connected) {
      throw StateError('Not connected');
    }
    if (_receiveQueue.isEmpty) {
      throw StateError('No messages in receive queue');
    }
    return _receiveQueue.removeFirst();
  }
}
